pandas/core/indexing.py

```diff
@@ -1838,7 +1855,16 @@ def _setitem_with_indexer_missing(self, indexer, value):
                     if len(value) != len(self.obj.columns):
                         raise ValueError("cannot set a row with mismatched columns")
 
-                value = Series(value, index=self.obj.columns, name=indexer)
+                if len(set(self.obj.dtypes)) > 1:
```

Comment:
this entire implementation should not be here, rather in Categorical indexing itself. Indexing is *already* very complicated and we are trying to remove things from core/indexing.py; this is dtype specific.

Reply:
pls address this comment

Reply:
I removed the dtype specific code from `core/indexing.py`. However, some code changes still remain there, because I couldn't find an alternative way. The latest changes address the issue and so the `Categorical` dtype is now preserved:
```python
import pandas as pd

df = pd.DataFrame.from_dict({'reg': [0,1,2], 'cat':pd.Categorical(['a','b','b'], categories=['a','b','c','d'])})
print(df.dtypes)  # reg is int64, cat is categorical

df.loc[3] = (3, 'c')  # add row with categorical value that exist in categories
print(df.dtypes)  # reg is int64, cat is now still categorical (which is not the case in master)
```

Some additional points related to this PR changes:
- When an unknown category is appended through index expansion in a dataframe then that value is turned to nan.
```python
In [1]: import pandas as pd
   ...: from pandas import Categorical
   ...: df = pd.DataFrame({"int": [0, 1], "cat": pd.Categorical(["a", "b"], categories=["a", "b"])})
   ...: df.loc[3] = [3.3, "d"]
   ...: df
Out[1]:
   int  cat
0  0.0    a
1  1.0    b
3  3.3  NaN
```
- The same issue is not addressed for `pd.Series` in this PR. Which means that the dtype of the series changes to object when an unseen category is appended through index expansion.
```python
In [1]: import pandas as pd
   ...: from pandas import Categorical
   ...: ser = pd.Series(Categorical(["a", "b", "c"]))
   ...: ser.loc[3] = "d"
   ...: ser
Out[1]: 
0    a
1    b
2    c
3    d
dtype: object
```